### ISSUE TITLE
angular.copy Very poor performance (large objects)

### ISSUE THREAD

**Issue by Tomino2112** [angular/angular.js#11099]
Our application loads resources from API and renders them in a table. API request takes ~200ms, rendering table ~140ms but when you actually looking at the page it takes about 6s to display it.

I have drilled through the code and realized this was caused by using angular.copy . I need to keep track of the old data so I copied the whole object to backup object via angular.copy() function. The copying itself took over 2s (~2500ms). 

I have searched for better ways, to copy objects and stumbled upon this code: 

```
cloneObject: function(o){
            const gdcc = "__getDeepCircularCopy__";
            if (o !== Object(o)) {
                return o; // primitive value
            }

            var set = gdcc in o,
                cache = o[gdcc],
                result;
            if (set && typeof cache == "function") {
                return cache();
            }
            // else
            o[gdcc] = function() { return result; }; // overwrite
            if (o instanceof Array) {
                result = [];
                for (var i=0; i<o.length; i++) {
                    result[i] = this.cloneObject(o[i]);
                }
            } else {
                result = {};
                for (var prop in o)
                    if (prop != gdcc)
                        result[prop] = this.cloneObject(o[prop]);
                    else if (set)
                        result[prop] = this.cloneObject(cache);
            }
            if (set) {
                o[gdcc] = cache; // reset
            } else {
                delete o[gdcc]; // unset again
            }
            return result;
        },
```
This is my implementation of this: http://stackoverflow.com/a/10729242/820942

Replacing all angular.copy() in my code with this **hugely** improved the webapp overall. The same example above takes ~41ms instead of ~2s. 

I should mention that the data I am mentioning is pretty large (renders into 10 000 cells table)

After very brief look on the angular.copy code I am aware of that there are some added angular goodies ($$hashkey) but still the difference in performance is just too big.

**Comment by pkozlowski-opensource** [comment]
@Tomino2112 did you try to put this code and see if it passes all the existing tests? As you've noticed, `angular.copy` is more of an internal utility and we kind of regret exposing it as a public API... but now the damage is done... So if your version works better for your particular case - by all means use it!

It is kind of hard to say more without an isolated reproduction scenario which we could profile to actually see what is going on. We are missing this as well as other essential info (ex.: browser and its version used for tests).

In short: please provide an isolated reproduce scenario, otherwise it is not very actionable.

**Comment by VitalyTomilov** [comment]
@Tomino2112, it may be interesting to see if your approach can be somehow used within angular for speed optimization, but in the meantime, it is never a good approach trying to display 10,000 cells on a page at once through a single request. It is a very poor approach from the architecture point of view.

@pkozlowski-opensource, I find his research credible enough to suggest that Angular's implementation for deep copy is to be reviewed, based on implementation by @Tomino2112. The performance difference seems to be huge.

**Comment by Tomino2112** [comment]
Hi guys, 
@pkozlowski-opensource When I have a minute I will create sample fiddle

@VitalyTomilov Yeah its not great to render that many cells, actually it is edge scenario, usually its around hundereds of cells - still not great though, but thats what you get when you try to render sheets of data online. I have some ideas for rendering on my list, just not enough time to play around with them at the moment. 

I will prepare some code example so you cna see it in action (i mean the angular.copy)

**Comment by jbedard** [comment]
I think the main thing slowing `angular.copy` is the use of a stack (arrays) to track the already copied objects.  The snippet proposed above modifies the objects being copied which might be faster but will have a lot of issues if you want a general purpose copy method (it won't work at all on immutable objects, what if it crashes in the middle leaving a bunch of objects with modifications, etc.).

#11215 helps a bit mainly by reducing the size of the stack (today sub object/arrays are pushed twice!).

However I have found that replacing the stack with an ES6 `Map` improves performance significantly. I'm not sure if it's worth shimming `Map` just for the copy method though. Would `Map` be useful anywhere else?

**Comment by gkalpak** [comment]
I wonder if a shimmed Map would have the same performance benefits as the "native" Map.
Isn't the private `HashMap` sort of a shim for `Map` ? Would it help ?

**Comment by jbedard** [comment]
`HashMap` modifies the key object so that won't work for this case. A shimmed `Map` has to use the 2-array/indexOf method so it wouldn't have any performance benefits.

**Comment by gkalpak** [comment]
That's what I thought :)

**Comment by VitalyTomilov** [comment]
Maybe it is a bit sinister having just one Copy version for both simple data and immutable objects, so if one wants a simpler, high-performance copy without any advanced provisions, it is not available.

If this is the case, the best approach is to provide two separate Copy implementations either through 2 separate functions or through an extra parameter for Copy.

**Comment by ladyborg** [comment]
Using something like this:

destination = JSON.parse(JSON.stringify(source));

I copied 10,000 key/value pairs of an object in ~5 milliseconds on my
machine. I know this leaves out a lot of the functionality found in
angular.copy, but perhaps there is some good way of merging this approach
in?

**Comment by jbedard** [comment]
`JSON.parse(JSON.stringify(source))` is often a good method if you know it can handle your data (no circles, no functions, no objects other then string/number/plain-object/null). A standard `copy` method could never make those assumptions though, and detecting such a thing would require traversing the entire source object which would kill any performance gain. Generally utility methods like this don't belong in angular anyway...

**Comment by ladyborg** [comment]
That's a good point, I didn't realize that code doesn't copy over functions or object methods :/

**Comment by realityking** [comment]
@gkalpak @jbedard How about using `Map` in newer browsers and falling back to a shim in old browsers (by my count Internet Explorer 9-10, Safari/iOS 6-7 and Android Browser)

**Comment by jbedard** [comment]
@realityking that's what I was proposing, I'm just not sure if creating that shim is worth one use case...

**Comment by gkalpak** [comment]
Considering how `copy()` is used in deep watchers and sometimes in `AST.prototype.primary()`, it might be worth the trouble. 

Two things I would take into account:

1. How much is the code overhead (in terms of loc).
2. How much is the performance benefit in "newer" browsers.

**Comment by jbedard** [comment]
When I tried it out (on top of #11215) I think it was ~50 LOC and made all modern browsers ~5x faster with my random over complicated test data.  I could open a PR if people think this is worth it?

**Comment by dpogue** [comment]
:+1: for using Map where supported

**Comment by lgalfaso** [comment]
@jbedard what about with not so complex data?

**Comment by jbedard** [comment]
As the complexity goes down (less objects) using `Map` has less benefits and is slower with simple data.  For example I found `copy({a:"b", c: {d: "e"}})` was about 30% slower using `Map`.

However in those simple data cases it is normally `isTypedArray(destination)` consuming the most time, not `new Map` or `Map.get/set`. So that 30% can easily be made up for with other changes such as https://github.com/jbedard/angular.js/commit/8a2503bdb5786dd31ebf503903406d7cbb686d5a which makes that simple copy 50% faster...

**Comment by ladyborg** [comment]
I don't have time for it right now, but if no one else does it by tomorrow
I can run some tests with varying sizes of each data-type being passed into
copy, both using Map and not using Map. Then we can pinpoint the number of
bytes in the median(s) of the valley(s) in the performance curve and
implement a check where the performance benefits of map outweigh the cons
(as @jbedard pointed out) and then copy can decide whether or not to use it.

What if we did something like this:

<pre style='color:#000000;background:#ffffff;'><html><body
style='color:#000000; background:#ffffff; '><pre>
<span style='color:#800000; font-weight:bold; '>var</span> byteThreshold
<span style='color:#808030; '>=</span> <span style='color:#696969; '>/*
Some number of bytes where objects</span>
<span style='color:#696969; '> of this size or greater are benefited</span>
<span style='color:#696969; '> by using map, and objects less than</span>
<span style='color:#696969; '> this byte size take a performance hit</span>
<span style='color:#696969; '> due to isTypedArray(destination).</span>
<span style='color:#696969; '> */</span><span style='color:#800080;
'>;</span>

<span style='color:#800000; font-weight:bold; '>if</span> <span
style='color:#808030; '>(</span>sizeOf<span style='color:#808030;
'>(</span><span style='color:#797997; '>source</span><span
style='color:#808030; '>)</span> <span style='color:#808030; '>></span>
byteThreshold<span style='color:#808030; '>)</span> <span
style='color:#800080; '>{</span>
<span style='color:#696969; '>// use map</span>
<span style='color:#800080; '>}</span> <span style='color:#800000;
font-weight:bold; '>else</span> <span style='color:#800080; '>{</span>
<span style='color:#696969; '>// don't use map</span>
<span style='color:#800080; '>}</span>
</pre>

And figure out where that threshold would be exactly? Or would it even be
worth it to check? Perhaps the performance tradeoffs are negligible? I
don't really know because I haven't run the tests or looked at the data,
but maybe jbedard@8a2503b
<https://github.com/jbedard/angular.js/commit/8a2503bdb5786dd31ebf503903406d7cbb686d5a>
already
knows where that would be?

**Comment by ladyborg** [comment]
I don't have time for it right now, but if no one else does it by tomorrow
I can run some tests with varying sizes of each data-type being passed into
copy, both using Map and not using Map. Then we can pinpoint the number of
bytes in the median(s) of the valley(s) in the performance curve and
implement a check where the performance benefits of map outweigh the cons
(as @jbedard pointed out) and then copy can decide whether or not to use it.

What if we did something like this:

<pre style='color:#000000;background:#ffffff;'><html><body
style='color:#000000; background:#ffffff; '><pre>
<span style='color:#800000; font-weight:bold; '>var</span> byteThreshold
<span style='color:#808030; '>=</span> <span style='color:#696969; '>/*
Some number of bytes where objects</span>
<span style='color:#696969; '> of this size or greater are benefited</span>
<span style='color:#696969; '> by using map, and objects less than</span>
<span style='color:#696969; '> this byte size take a performance hit</span>
<span style='color:#696969; '> due to isTypedArray(destination).</span>
<span style='color:#696969; '> */</span><span style='color:#800080;
'>;</span>

<span style='color:#800000; font-weight:bold; '>if</span> <span
style='color:#808030; '>(</span>sizeOf<span style='color:#808030;
'>(</span><span style='color:#797997; '>source</span><span
style='color:#808030; '>)</span> <span style='color:#808030; '>></span>
byteThreshold<span style='color:#808030; '>)</span> <span
style='color:#800080; '>{</span>
<span style='color:#696969; '>// use map</span>
<span style='color:#800080; '>}</span> <span style='color:#800000;
font-weight:bold; '>else</span> <span style='color:#800080; '>{</span>
<span style='color:#696969; '>// don't use map</span>
<span style='color:#800080; '>}</span>
</pre></pre>

And figure out where that threshold would be exactly? Or would it even be
worth it to check? Perhaps the performance tradeoffs are negligible? I
don't really know because I haven't run the tests or looked at the data,
but maybe jbedard@8a2503b already knows where that would be?

**Comment by ladyborg** [comment]
sorry, that code highlighter did not work out lol...

		var byteThreshold = /* Some number of bytes where objects
				of this size or greater are benefited
				by using map, and objects less than
				this byte size take a performance hit
				due to isTypedArray(destination).
				*/;

		if (sizeOf(source) > byteThreshold) {
			// use map
		} else {
			// don't use map
		}

**Comment by jbedard** [comment]
It is not worth having 2 implementations...

**Comment by Ledzz** [comment]
Had an issue: when I try to copy a large object, it copies in ~300ms, but sometimes it can copy for 18-20s. jbedard's method `JSON.parse(JSON.stringify(source))` worked good for me.

**Comment by jbedard** [comment]
33c67ce785cf8be7f0c294b3942ca4a337c5759d has improved this a bit more often making copy 1.5-3x faster.

**Comment by Tomino2112** [comment]
I have tried MANY solutions from all over the internet, but in the end I end-up writing specific function for copying my object. After all the tests and benchmark, the result is, if you have large arrays/objects, its better to write your own copy function specifically for that object with `for` loops because they are insanely fast compare to whatever other solutions. I still use angular's clone function, but only for small objects

**Comment by frfancha** [comment]
@jbedard Sorry for the naive question, but 33c67ce has a commit date June 9 and doesn't seem included in Angular 1.4.7, could you help me to understand version numbers?
Thanks

**Comment by jbedard** [comment]
It was just put into master a few days ago so it will be in the next 1.5 release.

**Comment by frfancha** [comment]
@jbedard Ok thanks for the info. We will try 1.5 then, I'm very curious about the performance difference. I hope there won't be too many breaking changes in 1.5

**Comment by stanleyxu2005** [comment]
I think angular should move these helper functions eg. `isNumber()`, `forEach()`, `copy`, `isEqual()` into angular-util package, so that the core package size can be even smaller, and we might be able to use lodash instead.

**Comment by gkalpak** [comment]
@stanleyxu2005, (as stated before) Angular never intended to provide a replacement for general purpose utility libraries. It's helper function (`copy()`, `forEach()`, `isXyz()` etc) where implemented for internal use (so they are focused on supporting/performing well only for the imtermally needed usecases - they might work well for other usecases, but it's not their purpose).

At that point, it seems like a good idea to expose the utility functions to the developers (since they were implemented anyway), so they didn't have to import a whole different library for simple usecases. This actually turned out to be a bad idea - among other things, it's much more difficult to make breaking changes to them to better support the evolving needs of the framework.

There's not much we can do now, because removing them break too many apps. The functions are also used internally, so removing them from the core is not an option (that's their purpose in the first place). So, they'll continue to exist, but users should keep in mind that they might not be the best option for their usecase.

**Comment by coli** [comment]
I just replaced angular.copy with JSON.parse(JSON.stringify()) on a 18meg tree structure. The speed difference was huge...

**Comment by bcherny** [comment]
One more case study.... 

```js
let as = [...] // array of ~4000 objects

console.time('angular')
angular.copy(as)
console.timeEnd('angular')
// => 71048.443ms

console.time('lodash')
_.cloneDeep(as)
console.timeEnd('lodash')
// => 521.026ms
```

Please deprecate the public angular.copy API, and add a note on the docs page!

**Comment by e-cloud** [comment]
The team should be aware that [source of `angular.copy`](https://github.com/angular/angular.js/blob/master/src/Angular.js#L811-L949) has too much nested functions, that's part of why it's so slow.

**Comment by gkalpak** [comment]
@e-cloud, is there any evidence that inlining the functions would improve performance?

**Comment by e-cloud** [comment]
Nope, what i mean is every time you call `angular.copy` will create nested functions, which cost more memory and time. Of source, in small scale it doesn't matter.

**Comment by stanleyxu2005** [comment]
@gkalpak Improve performance of frequently use functions is a common sense. I feel it is a bit confused of using angular functions. If those functions are intended to be used internally, better rename `foo` to `internal.foo` or `_foo`. If we don't have much knowledge of angular's history, we never know these functions are not for public usage.

**Comment by jbedard** [comment]
@stanleyxu2005 `JSON.parse(JSON.stringify(obj))` is very limited and does not work for the internal uses of `angular.copy`.

**Comment by gkalpak** [comment]
> every time you call `angular.copy` will create nested functions, which cost more memory and time.

@e-cloud, the question is whether the cost is neglible compared to other operations involved or not. If there is evidence that we can noticeably improve the performance of `angular.copy` (without sacrifizing functionality), we would be more than happy to do it.

**Comment by jbedard** [comment]
@e-cloud I tested your idea and it really makes no difference. It's only 3 closures per copy call. If it were 3 per object (and sub object) it would probably make a difference but that's not the case.

**Comment by e-cloud** [comment]
@jbedard , Do you try it in scale?

> Of source, in small scale it doesn't matter.

I've mention before, if only few calls.

**Comment by stripathix** [comment]
In my use case, I did not saw any a difference in angular.copy vs JSON.parse(JSON.stringfy(dataobject));

Rather _.map(dataobject, _.clone) was super fast.


**angular.copy(dataobject)**
2500ms

**JSON.parse(JSON.stringfy(dataobject))**
2500ms

**_.map(dataobject, _.clone)**
115ms
